### ISSUE TITLE
Add an option to change the order of choices

### ISSUE THREAD

**Issue by nicoTolly** [ulysseB/telamon#118]
Currently, the order of choices is not customizable, it is just provided by the api. For the sake of experimenting, we would like to have the possibility to alter this order, as we want to highlight the fact that this change can have a significant impact on the exploration time. 
In order to do so, we have to make the following modifications : 
- Modify telamon-gen to expose a list of all possible choices (we currently have a list of all actions, which are just instantiated choices - choices with a value). 
- modify config.rs to expose this possibility to the user.
- Maybe implement a shuffle or anything custom for the order of the choices we want to make.

Also, we want to allow a special king of exploration that would not necessarily go the leaves of the tree, but rather explore for example the 4 or 5 first levels exhaustively. This implies that we would not run anything on the device. These benchmarks could be launched with cargo bench.

**Comment by ulysseB** [comment]
Concerning the "special kind of exploration", you can create a function that does a BFS and stops at a fixed level without implementing the `Store` trait. I think the best would be to place this function directly in its `bench/` file. Maybe we will need to make some modules of `src/explorer` public for that, but I don't see it as a problem.

**Comment by nicoTolly** [comment]
As I understand it (and as discussed with Basile), an Action is a value taken on a specific choice. I think we just want to represent the choice. For example, for a given dimension dim1, a choice would be "choice of DimKind on dim1", whereas an Action would be something like "DimKind on dim1 is UNROLL".

**Comment by nicoTolly** [comment]
Maybe ActionEx does the job though ? I am checking that

**Comment by nicoTolly** [comment]
There is a struct Choice in telamon_gen/ir/choice.rs that corresponds exactly to what we need. I have the feeling that this structure is only exposed internally (as would suggest the name 'intermediate representation'). I think we should make use of that

**Comment by Elarnon** [comment]
@ulysseB We want to have the "Choice" (i.e. action without the domain value) because this is the old problem of "we want to be able to talk about a choice without materializing all the actions" (because otherwise we need to propagate everything). Also, I don't think we want to hardcode orders in the `choices::list` because that is highly non scalable if we want to explore a reasonable number of orderings. I think it should be very easy to switch to having the choice and then a lazy enumeration of the actual actions (we may want to avoid the naming confusion with what I call "choice" here and what we have in the explorer which is a list of `ActionEx`). Typically where we have in `explorer::choices::list` something like `gen_choice(orders.list(), &|o| Action::Order(lhs, rhs, o))` then we want to do something like `(Choice::Order(lhs, rhs), || gen_choice(orders.list(), &|o| Action::Order(lhs, rhs, o)))` and have both the Choice object and a lazy generator function and we can reorder the lazy generator function easily without having to actually propagate the constraints to generate the actual actions. This also is a format that can be much more easily used by the search algorithms  because they won't have to generate all the actions before selecting the order (with the caveat that some may be empty because fully specified ;) ) So maybe we don't want the first-class "Choice::Order(lhs, rhs)"? However it would be helpful while debugging to have an easy way to display a choice's name.

Actually we can just bundle the gen_choice call in a closure but then we really have *no* information about what the choice is and we can't be smart about the ordering we do.

@nicoTolly `ActionEx` is a wrapper around `Action` with some more data for lowering and tiling (iirc) which are currently not actions. The `telamon_gen::ir::choice` structure is internal to telamon-gen and we don't have access to this once the code is generated, so we can't use it directly. However it is used in the code generation of the `Action` class (see `telamon_gen/src/print/template/action.rs`; when we do `{# each choices #}` we iterate over values of type `telamon_gen::ir::choice::Choice` or equivalent AFAIK. Here what we need is to modify `telamon_gen/src/print/template/action.rs` and add a new `Choice` enum in addition to the `Action` enum which will be available in telamon proper (not telamon-gen). See below for a rough outline.

```rust
#[derive(PartialEq, Eq, Debug, Clone, Copy, Serialize, Deserialize)]
#[repr(C)]
pub enum Choice {
    {{~#each choices}}
        /// cbindgen:field-names=[{{~#each arguments}}{{this.[0]}}, {{/each~}}]
        {{to_type_name name}}(
        {{~#each arguments}}{{this.[1].def.keys.IdType}},{{/each~}}),
    {{/each~}}
}

impl From<Action> for Choice {
    fn from(action: Action) -> Choice {
        match action {
            {{~#each cohices}}
                Action::{{to_type_name name}}({{#each arguments}}{{this.[0]}}, {{/each}}_value) =>
                    Choice::{{to_type_name name}}({{#each arguments}}{{this.[0]}}, {{/each}})
        }
    }
}
```

**Comment by Elarnon** [comment]
@nicoTolly Is there anything more to do for this, or are we good after #141 ?

**Comment by nicoTolly** [comment]
It depends if we are satisfied with specifying only the order for the types of choice or if we want to allow a finer grain. I would say what we have is fine for the moment

**Comment by ulysseB** [comment]
I think I am going to close this for now. We can make a new issue if we have a more precise need.
